JuliaStream.jl/src/CUDAStream.jl

```diff
@@ -0,0 +1,146 @@
+include("Stream.jl")
+using CUDA
+
+const CuData = StreamData{T,CuArray{T}} where {T}
+const TBSize = 1024::Int
+const DotBlocks = 256::Int
+
+function devices()
+  return !CUDA.functional(false) ? [] :
+         map(d -> "$(CUDA.name(d)) ($(repr(d)))", CUDA.devices())
+end
+
+function blocks(data::CuData{T})::Int where {T}
+  return data.size ÷ TBSize
+end
+
+function make_stream(
+  arraysize::Int,
+  scalar::T,
+  device::Int,
+  silent::Bool,
+)::CuData{T} where {T}
+
+  if arraysize % TBSize != 0
+    error("arraysize ($(arraysize)) must be divisible by $(TBSize)!")
+  end
+
+  # so CUDA's device is 0 indexed, so -1 from Julia
+  CUDA.device!(device - 1)
+  selected = CUDA.device()
+  # show_reason is set to true here so it dumps CUDA info 
+  # for us regardless of whether it's functional
+  if !CUDA.functional(true)
+    error("Non-functional CUDA configuration")
+  end
+  data = CuData{T}(
+    CuArray{T}(undef, arraysize),
+    CuArray{T}(undef, arraysize),
+    CuArray{T}(undef, arraysize),
+    scalar,
+    arraysize,
+  )
+  if !silent
+    println("Using CUDA device: $(CUDA.name(selected)) ($(repr(selected)))")
+    println("Kernel parameters: <<<$(blocks(data)),$(TBSize)>>>")
+  end
+  return data
+end
+
+function init_arrays!(data::CuData{T}, init::Tuple{T,T,T}) where {T}
+  CUDA.fill!(data.a, init[1])
+  CUDA.fill!(data.b, init[2])
+  CUDA.fill!(data.c, init[3])
+end
+
+function copy!(data::CuData{T}) where {T}
+  function kernel(a, c)
+    i = (blockIdx().x - 1) * blockDim().x + threadIdx().x
```

Comment:
Same as with the AMDGPU one, worth checking the blocks and threads are indexed differently, adding a comment if necessary.

Reply:
Only blocks use 1-indexing, I'll add a comment in code. For reference, towards the end of the [section](https://cuda.juliagpu.org/stable/tutorials/introduction/#Writing-a-parallel-GPU-kernel), it used `index = (blockIdx().x - 1) * blockDim().x + threadIdx().x` for the global index.
This seems to be the preferred way of doing it, also seen in the [examples](https://github.com/JuliaGPU/CUDA.jl/blob/13dcce7415b7ed631208b00e901a6e6f76d11bf0/examples/vadd.jl#L6) and [internally](https://github.com/JuliaGPU/CUDA.jl/blob/13dcce7415b7ed631208b00e901a6e6f76d11bf0/src/mapreduce.jl#L144).

Reply:
Done